File: src/ts/src/ads/MultiBannerAd.ts
import {
    ObserverHandle,
    ObserverManager,
} from "../core";
import { AdObserver } from "./IAd";
import { IBannerAd } from "./IBannerAd";
import { IMultiAd } from "./IMultiAd";

export class MultiBannerAd extends ObserverManager<AdObserver> implements IBannerAd, IMultiAd<IBannerAd> {
    private _visible: boolean;
    private _anchor: [number, number];
    private _position: [number, number];
    private readonly _items: IBannerAd[];
    private readonly _handle: ObserverHandle;
    private _activeItem?: IBannerAd;
    private _loadedItems: IBannerAd[];

    public constructor() {
        super();
        this._anchor = [0, 0];
        this._position = [0, 0];
        this._visible = false;
        this._items = [];
        this._handle = new ObserverHandle();
        this._loadedItems = [];
    }

    public addItem(item: IBannerAd): void {
        this._items.push(item);
        item.isVisible = this._visible;
        this._handle.bind(item).addObserver({
            onLoaded: () => {
                if (this._loadedItems.indexOf(item) === -1) {
                    this._loadedItems.push(item);
                }
                this.invalidate();
                this.dispatchEvent(observer => observer.onLoaded && observer.onLoaded());
            },
            onClicked: () => this.dispatchEvent(observer => observer.onClicked && observer.onClicked()),
        });
    }

    public destroy(): void {
        for (const item of this._items) {
            item.destroy();
        }
        this._items.length = 0;
        this._handle.clear();
        this._activeItem = undefined;
    }

    public get isLoaded(): boolean {
        return this._items.some(item => item.isLoaded);
    }

    public async load(): Promise<boolean> {
        let result = false;
        for (const item of this._items) {
            if (item === this._activeItem && this._visible) {
                continue;
            }
            if (this._loadedItems.indexOf(item) !== -1) {
                continue;
            }
            if (await item.load()) {
                result = true;
            }
        }
        return true;
    }

    public get isVisible(): boolean {
        return this._visible;
    }

    public set isVisible(value: boolean) {
        this._visible = value;
        this.invalidate();
        for (const item of this._items) {
            if (this._loadedItems.indexOf(item) === -1) {
                // Load in background.ss
                const _ = item.load();
            }
        }
    }

    public get anchor(): [number, number] {
        return this._anchor;
    }

    public set anchor(value: [number, number]) {
        this._anchor = value;
        for (const item of this._items) {
            item.anchor = value;
        }
    }

    public get position(): [number, number] {
        return this._position;
    }

    public set position(value: [number, number]) {
        this._position = value;
        for (const item of this._items) {
            item.position = value;
        }
    }

    public get size(): [number, number] {
        let width = 0;
        let height = 0;
        for (const item of this._items) {
            const [itemWidth, itemHeight] = item.size;
            width = Math.max(width, itemWidth);
            height = Math.max(height, itemHeight);
        }
        return [width, height];
    }

    public set size(value: [number, number]) {
        for (const item of this._items) {
            item.size = value;
        }
    }

    private invalidate(): void {
        const lastActiveItem = this._activeItem;
        for (const item of this._items) {
            if (this._loadedItems.indexOf(item) !== -1) {
                this._activeItem = item;
                break;
            }
        }
        if (this._activeItem !== undefined) {
            this._activeItem.isVisible = this._visible;
        }
        if (lastActiveItem !== undefined && lastActiveItem !== this._activeItem) {
            lastActiveItem.isVisible = false;
        }
    }
}

File: src/unity/Runtime/Ads/MultiBannerAd.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;

namespace EE {
    public class MultiBannerAd : ObserverManager<AdObserver>, IBannerAd, IMultiAd<IBannerAd> {
        private bool _visible;
        private (float, float) _anchor;
        private (float, float) _position;
        private readonly List<IBannerAd> _items;
        private readonly ObserverHandle _handle;
        private IBannerAd _activeItem;
        private readonly HashSet<IBannerAd> _loadedItems;

        public MultiBannerAd() {
            _items = new List<IBannerAd>();
            _handle = new ObserverHandle();
            _loadedItems = new HashSet<IBannerAd>();
        }

        public void AddItem(IBannerAd item) {
            _items.Add(item);
            item.IsVisible = _visible;
            _handle.Bind(item).AddObserver(new AdObserver {
                OnLoaded = () => {
                    _loadedItems.Add(item);
                    Invalidate();
                    DispatchEvent(observer => observer.OnLoaded?.Invoke());
                },
                OnClicked = () => DispatchEvent(observer => observer.OnClicked?.Invoke())
            });
        }

        public void Destroy() {
            foreach (var item in _items) {
                item.Destroy();
            }
            _items.Clear();
            _handle.Clear();
            _activeItem = null;
            _loadedItems.Clear();
        }

        public bool IsLoaded => _items.Any(item => item.IsLoaded);

        public async Task<bool> Load() {
            var result = false;
            foreach (var item in _items) {
                if (item == _activeItem && _visible) {
                    // Ignore displaying item.
                    continue;
                }
                if (_loadedItems.Contains(item)) {
                    // Already loaded and not displayed.
                    continue;
                }
                if (await item.Load()) {
                    result = true;
                }
            }
            return result;
        }

        public bool IsVisible {
            get => _visible;
            set {
                _visible = value;
                Invalidate();
                foreach (var item in _items) {
                    if (!_loadedItems.Contains(item)) {
                        _ = item.Load();
                    }
                }
            }
        }

        public (float, float) Anchor {
            get => _anchor;
            set {
                _anchor = value;
                foreach (var item in _items) {
                    item.Anchor = value;
                }
            }
        }

        public (float, float) Position {
            get => _position;
            set {
                _position = value;
                foreach (var item in _items) {
                    item.Position = value;
                }
            }
        }

        public (float, float) Size {
            get {
                // Combined size of all ad views.
                var width = 0.0f;
                var height = 0.0f;
                foreach (var item in _items) {
                    var (itemWidth, itemHeight) = item.Size;
                    width = Math.Max(width, itemWidth);
                    height = Math.Max(height, itemHeight);
                }
                return (width, height);
            }
            set {
                foreach (var item in _items) {
                    item.Size = value;
                }
            }
        }

        private void Invalidate() {
            var lastActiveItem = _activeItem;
            foreach (var item in _items) {
                if (_loadedItems.Contains(item)) {
                    _activeItem = item;
                    break;
                }
            }
            if (_activeItem != null) {
                _activeItem.IsVisible = _visible;
            }
            if (lastActiveItem != null && lastActiveItem != _activeItem) {
                lastActiveItem.IsVisible = false;
            }
        }
    }
}